### ISSUE TITLE
RFC: Format Strings

### ISSUE THREAD

**Issue by hellerve** [carp-lang/Carp#1115]
As discussed in the Gitter chat, I recently implemented a very simple yet powerful version of format strings. This RFC is an effort to agree on a standard and push it into the standard library.

For fun and reference, here is the code in full:

```clojure
(hidden f-internal)
(defndynamic f-internal [s]
  (let [idx (String.index-of s \{)
        len (String.length s)]
    (if (= idx -1)
      [(list 'copy s)]
      (let [ss (String.suffix s (inc idx))
            endx (String.index-of ss \})]
        (if (= endx -1)
          (macro-error "error in format string: unclosed open bracket")
          (append [(list 'copy (String.prefix s idx)) (list 'str (parse (String.prefix ss endx)))]
            (f-internal (String.suffix ss (inc endx)))))))))

(doc f "formats a string using `str`.")
(defmacro f [s]
  (list 'String.concat (list 'ref (f-internal s))))
```

It expands into a single call to `String.concat`, with the string literal becoming copied chunks and the expressions evaluating in order and being stringified. Please note that this does allocate. I am not married to the name.

Here is how one would use it:

```clojure
(def x 1)
(def y "hi")

(IO.println &(f "this is x: {x} and this is the first letter of hi: {(head y)}"))
; => this is x: 1 and this is the first letter of hi: h
```

The ergonomics of this could be further improved by adding a reader macro, for instance using `f` in front of the string, like so:

```clojure
(def x 1)
(def y "hi")

(IO.println &f"this is x: {x} and this is the first letter of hi: {(head y)}")
; => this is x: 1 and this is the first letter of hi: h
```
Again, I’m not married to the glyph `f` here, but it seems like an obvious choice, and has prior art in Python f-strings, to which they are related.

I’d like to hear any and all comments, criticisms, or thoughts you might have!

Cheers

**Comment by scolsen** [comment]
This is great! I think `f` is fine as a reader macro. If we define it as a normal macro as well, I don't love the name `f` simply because it's typically used for local bindings for functions when writing in a functional style (f, g, h) etc. which might make it harder to decipher cases in which a locally bound lambda `f` is being called vs. the `f` macro. Since it uses `str`, would `fstr` be acceptable?

But again, as a reader macro, I think `f` is fine since it's unambiguous when appended to a string literal.

> Please note that this does allocate.

We should probably mention this in the doc for the macro.

Awesome work!

**Comment by hellerve** [comment]
Thanks for the kind words!

> Since it uses str, would fstr be acceptable?

I think that makes total sense, yes! If we have a reader macro I don’t think it will be used as a regular macro very often anyway; it’s also nice because having a literal form would make it clear that that’s how it’s supposed to work: while the macro also works with variables I think this gets brittle and ugly quickly.

**Comment by eriksvedang** [comment]
I like it a lot too. Could we use a more conventional sigil, like `$`?

`fstr` seems reasonable.

What should / could be removed in the existing core libs once this is in?

**Comment by hellerve** [comment]
> Could we use a more conventional sigil, like $?

As in, for the reader macro or for the interpolation? I followed the Python convention for both, but am happy to change either.

> What should / could be removed in the existing core libs once this is in?

I suppose a bunch of code could be simplified. I also think that this and `fmt` have some overlap, but each has distinct functionality that I think should be kept; what do y’all think?

**Comment by eriksvedang** [comment]
> As in, for the reader macro or for the interpolation? I followed the Python convention for both, but am happy to change either.

I meant the `f`. Oh I see, I'm not very familiar with Python. I think a non-letter prefix is nicer if possible – exactly which one doesn't really matter as long as it's typeable on most keyboards. The `{ }` inside the string is great.

**Comment by hellerve** [comment]
This is definitely possible—it does create a little friction with the `r` raw string sigil. But I think that ship has probably sailed.

**Comment by eriksvedang** [comment]
I forgot about that one... Do you mean that it's better to keep them all as letters for consistency? (I'm not sure which ship has sailed, sorry!)

**Comment by hellerve** [comment]
At  this point we have `r` for raw strings and `#` for patterns, so it’s too late to keep things consistent without breaking things. I’d happily change from `f` to `$` anyway, I just wanted to make sure we’re on the same page :)

**Comment by eriksvedang** [comment]
@hellerve Ah, right. Thanks!

**Comment by TimDeve** [comment]
I like this. What's the plan for escaping `{}`?

**Comment by hellerve** [comment]
In the version above, you can escape `{` with itself, meaning `{{` works. A closing bracket does not have any significance on its own, so I haven’t added an escape, but we probably should.

It’s also worth pointing out that putting a map literal inside a format string will confuse the format string because of the brackets. I don’t quite know how to resolve that, unless we rewrite this parser to respect nesting (i.e. counting the number of open brackets); but then I also don’t know how useful a map literal inside a format string would be.

**Comment by hellerve** [comment]
While we’re at it, I’d also like to point out another thing that bugs me a little, but is likely not resolvable: since format string literals are allocated, their type is `String` and not `(Ref String)`, leading to a bit of an incosistency. I think this is explainable in the docs, but it’s also unfortunate. Do you have any input on that?

**Comment by TimDeve** [comment]
Yeah I'm definitely aware of the potential confusion, especially the hidden alloc. On one hand the reader macro feels a bit more ergonomic but maybe keeping it a "normal" macro would be less confusing: `(template-string "x: {x}")`

**Comment by hellerve** [comment]
I think this can be  closed with #1142
